src/terminal/parser/OutputStateMachineEngine.cpp

```diff
@@ -1408,99 +1409,185 @@ static constexpr bool _isHexNumber(const wchar_t wch) noexcept
+try
 {
-    bool foundRGB = false;
+    bool foundRGBColorSpec = false;
     bool foundValidColorSpec = false;
+
+    bool isSharpSignFormat = false;
+    size_t rgbHexDigitCount = 0;
     std::array<unsigned int, 3> colorValues = { 0 };
+    std::array<unsigned int, 3> parameterValues = { 0 };
     bool success = false;
-    // We can have anywhere between [11,15] characters
-    // 9 "rgb:h/h/h"
-    // 12 "rgb:hh/hh/hh"
-    // Any fewer cannot be valid, and any more will be too many.
-    // Return early in this case.
-    //      We'll still have to bounds check when parsing the hh/hh/hh values
-    if (string.size() < 9 || string.size() > 12)
+    const auto stringSize = string.size();
+
+    // First we look for "rgb:"
+    // Other colorspaces are theoretically possible, but we don't support them.
+    auto curr = string.cbegin();
+    if (stringSize > 4)
     {
-        return false;
+        auto prefix = std::wstring(string.substr(0, 4));
+        std::transform(prefix.begin(), prefix.end(), prefix.begin(), ::towlower);
+        if (prefix.compare(L"rgb:") == 0)
+        {
+            // We can have one of the following formats:
+            // 9 "rgb:h/h/h"
+            // 12 "rgb:hh/hh/hh"
+            // 15 "rgb:hhh/hhh/hhh"
+            // 18 "rgb:hhhh/hhhh/hhhh"
+            // Any fewer cannot be valid, and any more will be too many.
+            // Return early in this case.
+            if (stringSize < 9 || stringSize > 18)
+            {
+                return false;
+            }
+
+            foundRGBColorSpec = true;
+
+            std::advance(curr, 4);
+        }
     }
 
-    // Now we look for "rgb:"
-    // Other colorspaces are theoretically possible, but we don't support them.
+    // Try the sharp sign format.
+    if (!foundRGBColorSpec && stringSize > 1)
+    {
+        const auto prefix = string.substr(0, 1);
+        if (prefix.compare(L"#") == 0)
+        {
+            // We can have one of the following formats:
+            // 4 "#hhh"
+            // 7 "#hhhhhh"
+            // 10 "#hhhhhhhhh"
+            // 13 "#hhhhhhhhhhhh"
+            // Any other cases will be invalid.
+            // Return early in this case.
+            if (!(stringSize == 4 || stringSize == 7 || stringSize == 10 || stringSize == 13))
+            {
+                return false;
+            }
 
-    auto curr = string.cbegin();
-    if ((*curr++ == L'r') &&
-        (*curr++ == L'g') &&
-        (*curr++ == L'b') &&
-        (*curr++ == L':'))
+            isSharpSignFormat = true;
+            foundRGBColorSpec = true;
+            rgbHexDigitCount = (stringSize - 1) / 3;
+
+            std::advance(curr, 1);
+        }
+    }
+
+    // Try the XOrg app color name.
+    if (!foundRGBColorSpec)
     {
-        foundRGB = true;
+        try
+        {
+            til::color color = Utils::ColorFromXOrgAppColorName(string);
+            rgb = RGB(color.r, color.g, color.b);

```

Comment:
Note that `til::color` will automatically convert to a `COLORREF`, so this could just be:
````
rgb = Utils::ColorFromXOrgAppColorName(string);
````
That said, why can't the color map just use `COLORREF` entries to start with? Using `til::color` here is just making the whole process more memory intensive and less efficient. I don't see the benefit.

Reply:
til::color _should_ take up as many bytes as a COLORREF, but support alpha (RGBA instead of 0BGR). If it’s not doing so, we should absolutely fix that.

Reply:
(And pulling out individual components versus byte-swapping it into a COLORREF should be a wash, right? If not, it is perhaps reasonable to pull the whole thing!)

Reply:
Clever: clang will optimize a conversion from r,g,b,a into COLORREF by masking off the alpha (presumably only in little endian!), but MSVC cannot see through it and emits a few byte swaps totalling 7 instructions.

Reply:
Yeah sorry, I was mistaken about the additional memory. But there's definitely unnecessary overhead from all that byte shifting. I've actually tried messing with the `til::color` class before, and I couldn't get it to cleanly handle round tripping of a COLORREF. Again it's probably not that big a deal - it just irks me that we're introducing overhead for no real reason.

Reply:
That’s totally reasonable. I was under the mistaken impression (until just now!) that it was nigh-unto free, and so have been liberal with it.

Reply:
This is keeping me awake, but I can’t help it.

It is skirting the rules of defined behavior, but do you think there is a place for a (little endian only) implementation of the COLORREF conversion operators that is a memcpy and a mask? It wouldn’t be dissimilar from treating the body of til::color as a union of four u8s and a single u32.

An optimizing compiler would almost certainly lower a memcpy of four bytes as a register loading operation, and the mask is a single instruction on top of that.

Reply:
> It is skirting the rules of defined behavior, but do you think there is a place for a (little endian only) implementation of the COLORREF conversion operators that is a memcpy and a mask?

If that's what it takes to make the code optimize reasonably then I'd be inclined to do it. Throw in some static asserts or whatever to try and make sure it's safe, and maybe one day we can replace it with something cleaner if the MSVC optimizer improves. Otherwise I'm afraid the overhead tends to make me avoid a lot of these helper classes. And I know I probably worry too much about things like this (in this particular case it's really not that significant), but I'm from an era when it was normal to obsess over every byte of code generated. When I look at the assembly we're generating in some places it makes me want to weep.